base_layer/wallet/src/storage/sqlite_db.rs

```diff
@@ -27,27 +27,175 @@ use crate::{
 };
 use diesel::{prelude::*, result::Error as DieselError, SqliteConnection};
 
+use crate::util::encryption::{decrypt_bytes_integral_nonce, encrypt_bytes_integral_nonce, AES_NONCE_BYTES};
+use aes_gcm::{
+    aead::{generic_array::GenericArray, Aead},
+    Aes256Gcm,
+};
+use log::*;
 use std::{
     convert::TryFrom,
-    sync::{Arc, Mutex},
+    sync::{Arc, Mutex, RwLock},
+};
+use tari_comms::{
+    peer_manager::Peer,
+    types::{CommsPublicKey, CommsSecretKey},
+};
+use tari_crypto::{
+    keys::PublicKey,
+    tari_utilities::{
+        hex::{from_hex, Hex},
+        ByteArray,
+    },
 };
-use tari_comms::{peer_manager::Peer, types::CommsSecretKey};
-use tari_crypto::tari_utilities::{hex::Hex, ByteArray};
+
+const LOG_TARGET: &str = "wallet::storage::wallet_sqlite_database";
 
 /// A Sqlite backend for the Output Manager Service. The Backend is accessed via a connection pool to the Sqlite file.
 #[derive(Clone)]
 pub struct WalletSqliteDatabase {
     database_connection: Arc<Mutex<SqliteConnection>>,
+    cipher: Arc<RwLock<Option<Aes256Gcm>>>,
 }
 impl WalletSqliteDatabase {
-    pub fn new(database_connection: Arc<Mutex<SqliteConnection>>) -> Self {
-        Self { database_connection }
+    pub fn new(
+        database_connection: Arc<Mutex<SqliteConnection>>,
+        cipher: Option<Aes256Gcm>,
+    ) -> Result<Self, WalletStorageError>
+    {
+        // Here we validate if the database is encrypted or not and if a cipher is provided that it is the correct one.
+        // Unencrypted the database should contain a CommsPrivateKey and associated CommsPublicKey
+        // Encrypted the data should contain a CommsPublicKey in the clear and an encrypted CommsPrivateKey
+        // To confirm if the provided Cipher is correct we decrypt the CommsPrivateKey and see if it produces the same
+        // CommsPublicKey that is stored in the db
+        {
+            let conn = acquire_lock!(database_connection);
+            let secret_key = WalletSettingSql::get(format!("{}", DbKey::CommsSecretKey), &conn)?;
+            let db_public_key = WalletSettingSql::get(format!("{}", DbKey::CommsPublicKey), &conn)?;
```

Comment:
nit: using `to_string` instead of `format!` is more idiomatic
```suggestion
            let secret_key = WalletSettingSql::get(DbKey::CommsSecretKey.to_string()), &conn)?;
            let db_public_key = WalletSettingSql::get(DbKey::CommsPublicKey.to_string()), &conn)?;
```

Reply:
If you wanted to get fancy:

```rust
   pub fn get<T: ToString>(key: T, conn: &SqliteConnection) -> Result<Option<String>, WalletStorageError> {
        wallet_settings::table
            .filter(wallet_settings::key.eq(key.to_string()))
           (...)
    }
```
and then

```rust
let secret_key = WalletSettingSql::get(DbKey::CommsSecretKey), &conn)?;
```

Reply:
though, you probably just want `WalletSettingsSql` to only take in a `DbKey`. This way an invalid key will produce a compile error.

Reply:
I am not sure I like the DbKey leaking from the WalletBackend trait implementation. WalletSettingSql and all the *Sql structs are direct views into the DB really. The logic of how the db is being used and its input/output validated live in the trait implementation and the WalletSqliteDatabase.

Reply:
It's all in the same `storage` module and internal to WalletBackend, so not sure I'd call that a leak. In any case, I think the code is neater if you use either DbKey directly or the `T: ToString`

Reply:
I disagree about it being a leak because if we wanted to just nix the whole WalletBackend trait abstraction for whatever reason we would have to come into this specialized Db interface struct and make a change.